src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/ExtractToCodeBehindCodeActionProvider.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Components;
+using Microsoft.AspNetCore.Razor.Language.Extensions;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.CodeAnalysis;
+using Newtonsoft.Json.Linq;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Microsoft.AspNetCore.Razor.Language.Legacy;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Refactoring
+{
+    internal class ExtractToCodeBehindCodeActionProvider : RazorCodeActionProvider
+    {
+        override public Task<CommandOrCodeActionContainer> ProvideAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+        {
+            if (context.Document.IsUnsupported())
+            {
+                return Task.FromResult<CommandOrCodeActionContainer>(null);
+            }
+
+            if (!FileKinds.IsComponent(context.Document.GetFileKind()))
+            {
+                return Task.FromResult<CommandOrCodeActionContainer>(null);
+            }
+
+            var change = new SourceChange(context.Location.AbsoluteIndex, length: 0, newText: string.Empty);
+            var node = context.Document.GetSyntaxTree().Root.LocateOwner(change);
+            if (node is null)
+            {
+                return Task.FromResult<CommandOrCodeActionContainer>(null);
+            }
+
+            while (!(node is RazorDirectiveSyntax))
+            {
+                node = node.Parent;
+                if (node == null)
+                {
+                    return Task.FromResult<CommandOrCodeActionContainer>(null);
+                }
+            }
+
+            if (!(node is RazorDirectiveSyntax))
+            {
+                return Task.FromResult<CommandOrCodeActionContainer>(null);
+            }
+            var directiveNode = (RazorDirectiveSyntax)node;
+
+            if (directiveNode.DirectiveDescriptor != ComponentCodeDirective.Directive && directiveNode.DirectiveDescriptor != FunctionsDirective.Directive)
```

Comment:
Unfortunately, every few conditions we modify or change variables we care about. I could still see doing this but it would amount to cutting the code verbatim into another method I think.